### ISSUE TITLE
require 'openssl' causes NameError (uninitialized constant OpenSSL) in 1.0.0 RC14-dev

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1614]
Version:

```
truffleruby 1.0.0-rc14-dev-f6b387c4, like ruby 2.6.1, Interpreted [x86_64-darwin]
```

To reproduce:

```ruby
require 'openssl'
```

Error:

```
Traceback (most recent call last):
       13: from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/bin/irb:31:in `<main>'
       12: from (irb):2
       11: from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/rubygems/core_ext/kernel_require.rb:54:in `require'
       10: from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/rubygems/core_ext/kernel_require.rb:54:in `require'
        9: from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/openssl.rb:13:in `<top (required)>'
        8: from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/rubygems/core_ext/kernel_require.rb:54:in `require'
        7: from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/rubygems/core_ext/kernel_require.rb:54:in `require'
        6: from ossl.c:1130:in `Init_openssl'
        5: from ruby.c:2096:in `rb_define_module'
        4: from ruby.c:2100:in `rb_define_module_under'
        3: from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/truffle/truffle/cext.rb:1188:in `rb_define_module_under'
        2: from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/truffle/truffle/cext.rb:1188:in `const_get'
        1: from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/truffle/truffle/cext.rb:1188:in `const_missing'
NameError (uninitialized constant OpenSSL)
```

**Comment by deepj** [comment]
@chrisseaton another regression

**Comment by chrisseaton** [comment]
I can't reproduce that. Is that file literally what you're running?

**Comment by chrisseaton** [comment]
That's not even our backtrace format. Can you show a full terminal session?

**Comment by nirvdrum** [comment]
@deepj It looks like you're tracking TruffleRuby master. If you've been doing so since before the Ruby 2.6.1 merge, you need to the extensions for the built-in native libraries. I'd recommend just doing a `git clean -dxf` and rebuilding everything. Please let me know if that clears up your issue.

**Comment by deepj** [comment]
This is strange. I can't reproduce the above way in a different terminal session. But I can reproduce it in my specs.

@chrisseaton I can offer a private repo here on Github. It's some old assignment which I upgraded to Ruby 2.6.1 and I tested some stuff in TruffleRuby

@nirvdrum this is a nightly build offered by @chrisseaton to me

**Comment by chrisseaton** [comment]
Are you perhaps running a different Ruby as a sub-process?

**Comment by deepj** [comment]
I use chruby. In the direction I have `.ruby-version` set to 2.6.1.

Installed rubies
```
   ruby-2.5.3
   ruby-2.6.1
*  truffleruby-1.0.0-nightly
   truffleruby-1.0.0-rc11
   truffleruby-1.0.0-rc12
   truffleruby-1.0.0-rc13
```

**Comment by deepj** [comment]
This is an output when I run specs

```
$ rspec

An error occurred while loading spec_helper.
Failure/Error: require_relative '../app'

NameError:
  uninitialized constant OpenSSL
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/truffle/truffle/cext.rb:1188:in `const_missing'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/truffle/truffle/cext.rb:1188:in `const_get'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/truffle/truffle/cext.rb:1188:in `rb_define_module_under'
# ruby.c:2100:in `rb_define_module_under'
# ruby.c:2096:in `rb_define_module'
# ossl.c:1130:in `Init_openssl'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/openssl.rb:13:in `require'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/openssl.rb:13:in `<top (required)>'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/net/https.rb:23:in `require'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/net/https.rb:23:in `<top (required)>'
# ~/.gem/truffleruby/2.6.1/gems/down-4.8.0/lib/down/net_http.rb:4:in `require'
# ~/.gem/truffleruby/2.6.1/gems/down-4.8.0/lib/down/net_http.rb:4:in `<top (required)>'
# ~/.gem/truffleruby/2.6.1/gems/down-4.8.0/lib/down.rb:4:in `require'
# ~/.gem/truffleruby/2.6.1/gems/down-4.8.0/lib/down.rb:4:in `<top (required)>'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/bundler/runtime.rb:81:in `require'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/bundler/runtime.rb:81:in `block (2 levels) in require'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/bundler/runtime.rb:76:in `each'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/bundler/runtime.rb:76:in `block in require'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/bundler/runtime.rb:65:in `each'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/bundler/runtime.rb:65:in `require'
# ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/bundler.rb:114:in `require'
# ./boot.rb:10:in `<top (required)>'
# ./app.rb:3:in `require_relative'
# ./app.rb:3:in `<top (required)>'
# ./spec/spec_helper.rb:5:in `require_relative'
# ./spec/spec_helper.rb:5:in `<top (required)>'

An error occurred while loading ./spec/requests/exchange_spec.rb.
Failure/Error: RSpec.describe App, type: :request do

NameError:
  uninitialized constant App
# ./spec/requests/exchange_spec.rb:5:in `const_missing'
# ./spec/requests/exchange_spec.rb:5:in `<top (required)>'


Finished in 0.011 seconds (files took 14 seconds to load)
0 examples, 0 failures, 2 errors occurred outside of examples
```

**Comment by deepj** [comment]
@chrisseaton would you like have an access to the repo to try to reproduce it?

**Comment by deepj** [comment]
@chrisseaton The issue is easily reproducible using this

```
gem install skylight
```

Error
```
Fetching skylight-core-3.1.4.gem
Fetching skylight-3.1.4.gem
Successfully installed skylight-core-3.1.4
Building native extensions. This could take a while...
ERROR:  Error installing skylight:
	ERROR: Failed to build gem native extension.

    current directory: ~/.gem/truffleruby/2.6.1/gems/skylight-3.1.4/ext
~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/bin/truffleruby -I ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri -r ./siteconf20190312-46381-k9gr7d.rb extconf.rb
~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/truffle/truffle/cext.rb:1188:in `const_missing': uninitialized constant OpenSSL (NameError)
	from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/truffle/truffle/cext.rb:1188:in `const_get'
	from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/truffle/truffle/cext.rb:1188:in `rb_define_module_under'
	from ruby.c:2100:in `rb_define_module_under'
	from ruby.c:2096:in `rb_define_module'
	from ossl.c:1130:in `Init_openssl'
	from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/rubygems/core_ext/kernel_require.rb:117:in `require'
	from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/rubygems/core_ext/kernel_require.rb:117:in `require'
	from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/openssl.rb:13:in `<top (required)>'
	from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/rubygems/core_ext/kernel_require.rb:117:in `require'
	from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/rubygems/core_ext/kernel_require.rb:117:in `require'
	from ~/.gem/truffleruby/2.6.1/gems/skylight-3.1.4/lib/skylight/util/ssl.rb:1:in `<top (required)>'
	from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/rubygems/core_ext/kernel_require.rb:117:in `require'
	from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/rubygems/core_ext/kernel_require.rb:117:in `require'
	from ~/.gem/truffleruby/2.6.1/gems/skylight-3.1.4/lib/skylight/native_ext_fetcher.rb:7:in `<top (required)>'
	from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/rubygems/core_ext/kernel_require.rb:117:in `require'
	from ~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/lib/mri/rubygems/core_ext/kernel_require.rb:117:in `require'
	from extconf.rb:20:in `<main>'
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--without-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=~/.rubies/truffleruby-1.0.0-nightly/jre/languages/ruby/bin/truffleruby

extconf failed, exit code 1

Gem files will remain installed in ~/.gem/truffleruby/2.6.1/gems/skylight-3.1.4 for inspection.
Results logged to ~/.gem/truffleruby/2.6.1/extensions/x86_64-darwin/2.6.0/skylight-3.1.4/gem_make.out
```

**Comment by chrisseaton** [comment]
I think I've now seen something similar locally now, but it's a very nonsensical problem. I mean if we're installing gems then `openssl` is definitely there. Maybe it's being required in a slightly unusual way.

**Comment by deepj** [comment]
@chrisseaton This didn't happen in RC13 ...

**Comment by chrisseaton** [comment]
Yes I know - it's a bug, it's not intended.

**Comment by chrisseaton** [comment]
This appears to be some kind of autoload issue.

**Comment by chrisseaton** [comment]
I still don't know exactly why it goes wrong but here's two points of it.

```
[ruby] INFO: /Users/chrisseaton/Documents/truffleruby/truffleruby/mxbuild/graalvm/jre/languages/ruby/lib/mri/net/http.rb:25: setting up autoload Object::OpenSSL with openssl
....
/Users/chrisseaton/Documents/truffleruby/truffleruby/mxbuild/graalvm/jre/languages/ruby/lib/truffle/truffle/cext.rb:1188:in `const_missing': uninitialized constant OpenSSL (NameError)
```

**Comment by chrisseaton** [comment]
Reproduced by

```ruby
require 'net/http'
require 'openssl'
```

**Comment by chrisseaton** [comment]
Found the issue - thinking about a fix.

**Comment by chrisseaton** [comment]
`test.rb`:

```ruby
autoload :Test2, './test2.rb'
require_relative './test2.rb'
```

`test2.rb`

```ruby
p Object.autoload?(:Test2)
p Object.const_defined?(:Test2, false)
p Object.const_get(:Test2, false)
```

```
$ ruby test.rb 
nil
false
Traceback (most recent call last):
	3: from test.rb:2:in `<main>'
	2: from test.rb:2:in `require_relative'
	1: from /Users/chrisseaton/Documents/truffleruby/truffleruby/test2.rb:3:in `<top (required)>'
/Users/chrisseaton/Documents/truffleruby/truffleruby/test2.rb:3:in `const_get': uninitialized constant Test2 (NameError)
```

```
$ truffleruby-1.0.0-rc13-macos-amd64/bin/ruby test.rb 
"./test2.rb"
true
/Users/chrisseaton/Documents/truffleruby/truffleruby/test2.rb:3:in `const_missing': uninitialized constant Test2 (NameError)
Did you mean?  Test2
	from /Users/chrisseaton/Documents/truffleruby/truffleruby/test2.rb:3:in `const_get'
	from /Users/chrisseaton/Documents/truffleruby/truffleruby/test2.rb:3:in `<top (required)>'
	from test.rb:2:in `require_relative'
	from test.rb:2:in `<main>'
```

```
$ jt ruby test.rb 
"./test2.rb"
true
/Users/chrisseaton/Documents/truffleruby/truffleruby/src/main/ruby/core/module.rb:102:in `const_missing': uninitialized constant Test2 (NameError)
	from /Users/chrisseaton/Documents/truffleruby/truffleruby/src/main/ruby/core/module.rb:102:in `const_get'
	from /Users/chrisseaton/Documents/truffleruby/truffleruby/test2.rb:3:in `<top (required)>'
	from test.rb:2:in `require_relative'
	from test.rb:2:in `<main>'
```

I don't think it's a regression. Maybe something else changed and exposed this bug.

**Comment by eregon** [comment]
Yes, what changed is that `lib/mri/net/http.rb` now autoloads OpenSSL instead of `require`-ing it directly.

**Comment by eregon** [comment]
Actually the diff is:
```diff
diff --git a/lib/mri/net/http.rb b/lib/mri/net/http.rb
index 90f051978f..31cbb7a866 100644
--- a/lib/mri/net/http.rb
+++ b/lib/mri/net/http.rb
@@ -20,11 +20,11 @@
 # See Net::HTTP for an overview and examples.
 #

-require 'net/protocol'
+require_relative 'protocol'
 require 'uri'
+autoload :OpenSSL, 'openssl'

 module Net   #:nodoc:
-  autoload :OpenSSL, 'openssl'
```

So the `autoload` moved.

**Comment by chrisseaton** [comment]
Turns out this was tagged all along

```
fails:Module#autoload the autoload is removed when the same file is required directly without autoload with a full path
fails:Module#autoload the autoload is removed when the same file is required directly without autoload with a relative path
fails:Module#autoload the autoload is removed when the same file is required directly without autoload in a nested require
```

**Comment by chrisseaton** [comment]
Workaround on the way.

**Comment by chrisseaton** [comment]
Worked around in 3a8b2c204d394b7c9d4909a5b76407409af17f86.

**Comment by eregon** [comment]
I added a proper fix for this in ec974a98b905f880b3455af854068027f60b6f3c.
